Building: Bad interface file/Cannot continue after interface file error

Hi, I'm coming from Arch Linux and working on getting GHC 9.4 running. The rebuilds require `happy` to compile.

Building `happy` 2.0.2 with GHC 9.4.8 leads to this error:
```
Declaration for lalrBackendDataDir
Unfolding of lalrBackendDataDir:
  Bad interface file: /usr/lib/ghc-9.4.8/site-local/happy-lib-2.0.2/Paths_happy_lib.dyn_hi
      Something is amiss; requested module  happy-lib-2.0.2-6uhwNS4KElfEeCW2LxXwlE-backend-lalr:Paths_happy_lib differs from name found in the interface file happy-lib-2.0.2-CsmyxCqWSYG1DawVd57fhe-backend-glr:Paths_happy_lib (if these names look the same, try again with -dppr-debug)

<no location info>: error:
    Cannot continue after interface file error
```

### Investigation
The same happens with `happy` 2.1.3. Superficially, the error looks similar to #315. One difference is that, for bootstrapping reasons, we use the GHC included `runhaskell` script instead of `cabal install`.

Building with GHC 9.2 and Cabal 3.6 emits the same `Bad interface file [...]` message, but does not fail to build. See error log below for details.

### Conclusion
For the time being, is there anything we can do to workaround this issue?

If you need anything else from me or want me to check something, feel free to ask.
-- Vekhir.

## Appendix
### System information:
OS: Arch Linux
Kernel: Linux 6.12.1-arch1-1
GHC: 9.4.8
Cabal: 3.8.1.0
happy-lib: 2.0.2 and 2.1.3 ([PKGBUILD](https://gitlab.archlinux.org/archlinux/packaging/packages/haskell-happy-lib/-/blob/25ce4245d2672b11e0d5db362b19de4d093667fa/PKGBUILD))
happy: 2.0.2 and 2.1.3 ([PKGBUILD](https://gitlab.archlinux.org/archlinux/packaging/packages/happy/-/blob/9284b490da702a934a847be61462a4d40d010251/PKGBUILD))

### Error log
GHC 9.4:

- [happy-2.0.2-1-x86_64-build.log](https://github.com/user-attachments/files/18009800/happy-2.0.2-1-x86_64-build.log)
- [happy-2.1.3-1-x86_64-build.log](https://github.com/user-attachments/files/18009980/happy-2.1.3-1-x86_64-build.log)

GHC 9.2:
- [happy-2.0.2-1-x86_64-build.log](https://github.com/user-attachments/files/18010096/happy-2.0.2-1-x86_64-build.log)

OK, I turned your example into this PKGBUILD: 
[PKGBUILD.txt](https://github.com/user-attachments/files/18239311/PKGBUILD.txt)

It clones the github repo and switches to it, then executes the instructions as above. I had to add `--disable-library-vanilla`, as otherwise I'd get this error:
```
Configuring happy-lib-2.1.3...
Preprocessing library 'grammar' for happy-lib-2.1.3..
Building library 'grammar' for happy-lib-2.1.3..
[1 of 2] Compiling Happy.Grammar.ExpressionWithHole ( grammar/src/Happy/Grammar/ExpressionWithHole.hs, dist/build/grammar/Happy/Grammar/ExpressionWithHole.o, dist/build/grammar/Happy/Grammar/ExpressionWithHole.dyn_o )

grammar/src/Happy/Grammar/ExpressionWithHole.hs:1:8: error:
    Could not find module ‘Prelude’
    There are files missing in the ‘base-4.16.4.0’ package,
    try running 'ghc-pkg check'.
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
1 | module Happy.Grammar.ExpressionWithHole where
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
With that, building `happy-lib` succeeds, but then this error occurs:
```
ghc-9.2.8: can't find a package database at lib/dist/package.conf.inplace
```
Well, we need to switch back to the top-level :)

Interestingly, there is no error message at all - previously, it would complain, but nothing happens. I then confirmed with GHC 9.6.6 and it actually works!
Not quite sure what's going on. We usually install the packages via `register.sh`, maybe something there is going off?

[Full build log (GHC 9.2.8)](https://github.com/user-attachments/files/18239319/happy-test-2.0.2-1-x86_64-build.log)
[register.sh for happy-lib](https://github.com/user-attachments/files/18239378/register_happy_lib.txt)

--- a/lib/backend-glr/src/Happy/Backend/GLR.hs
+++ b/lib/backend-glr/src/Happy/Backend/GLR.hs
@@ -1,6 +1,6 @@
 module Happy.Backend.GLR where
 
-import Paths_happy_lib
+import Happy.Paths
 
 glrBackendDataDir :: IO String
 glrBackendDataDir = getDataDir

--- a/lib/backend-glr/src/Happy/Backend/GLR/ProduceCode.lhs
+++ b/lib/backend-glr/src/Happy/Backend/GLR/ProduceCode.lhs
@@ -17,7 +17,7 @@ This module is designed as an extension to the Haskell parser generator Happy.
 >                       , Options
 >                       ) where
 
-> import Paths_happy_lib ( version )
+> import Happy.Paths ( version )
 > import Happy.Grammar
 > import Happy.Grammar.ExpressionWithHole ( substExpressionWithHole )
 > import Happy.Tabular.LALR

--- a/lib/backend-lalr/src/Happy/Backend/LALR.hs
+++ b/lib/backend-lalr/src/Happy/Backend/LALR.hs
@@ -1,6 +1,6 @@
 module Happy.Backend.LALR where
 
-import Paths_happy_lib
+import Happy.Paths
 import Data.Char
 
 lalrBackendDataDir :: IO String

--- a/lib/backend-lalr/src/Happy/Backend/LALR/ProduceCode.lhs
+++ b/lib/backend-lalr/src/Happy/Backend/LALR/ProduceCode.lhs
@@ -6,7 +6,7 @@ The code generator.
 
 > module Happy.Backend.LALR.ProduceCode (produceParser) where
 
-> import Paths_happy_lib  ( version )
+> import Happy.Paths  ( version )
 > import Data.Version              ( showVersion )
 > import Happy.Grammar
 > import Happy.Grammar.ExpressionWithHole ( substExpressionWithHole )

--- a/lib/happy-lib.cabal
+++ b/lib/happy-lib.cabal
@@ -106,7 +106,23 @@ library tabular
                        Happy.Tabular.First,
                        Happy.Tabular.Info,
                        Happy.Tabular.LALR,
-                       Happy.Tabular.NameSet
+                       Happy.Tabular.NameSet,
+  -- The following 3 lines constitute a workaround for cabal#10687, triggered by #328.
+  -- The setup in #328 triggers a clashing use of Paths_happy_lib between
+  -- `backend-lalr` and `backend-glr` in the presence of --libsubdir.
+  -- The workaround moves Paths_happy_lib into `tabular` where it can be shared.
+  -- However, in order to really refer to the Paths_happy_lib from `tabular`,
+  -- we have to reexport it under a different name, because otherwise we get
+  --
+  --    These modules are needed for compilation but not listed in your .cabal file's other-modules for ‘happy-lib-2.1.3-inplace-backend-lalr’ :
+  --        Paths_happy_libsuppress
+  --
+                       Paths_happy_lib
+  reexported-modules:  Paths_happy_lib as Happy.Paths
+  autogen-modules:     Paths_happy_lib
+  -- end of workaround. Delete the workaround and use the bash script posted at
+  -- https://github.com/haskell/happy/issues/328#issuecomment-2597598320
+  -- to test whether the workaround is still needed.
   build-depends:       base < 5, array, containers, grammar
 
 
@@ -116,8 +132,6 @@ library backend-lalr
   exposed-modules:     Happy.Backend.LALR,
                        Happy.Backend.LALR.ProduceCode
   build-depends:       base < 5, array, grammar, tabular
-  other-modules:       Paths_happy_lib
-  autogen-modules:     Paths_happy_lib
 
 library backend-glr
   import: common-stanza
@@ -125,8 +139,6 @@ library backend-glr
   exposed-modules:     Happy.Backend.GLR,
                        Happy.Backend.GLR.ProduceCode
   build-depends:       base < 5, array, grammar, tabular
-  other-modules:       Paths_happy_lib
-  autogen-modules:     Paths_happy_lib
 
 library
   import: common-stanza

